### ISSUE TITLE
[GPU][BF16] torch.mul is lowered to hlo as f32 multiply

### ISSUE THREAD

**Issue by apivovarov** [pytorch/xla#8545]
## ❓ Questions and Help

torch 2.5.1
torch_xla 2.5.1
cuda 12.4
GPU NVIDIA L4

The following example uses `torch.mul` where both operands are bf16, but in the HLO graph, I see an f32 multiply operation.

```
export XLA_FLAGS="--xla_dump_to=/tmp/dump --xla_dump_hlo_as_text --xla_dump_hlo_pass_re=.*"
```
```
import torch
import torch_xla as xla

device = xla.device(0)

def foo(a, b):
  y = torch.mul(a, b)
  return y

a = torch.ones([5, 9216, 64], dtype=torch.bfloat16, device=device)
b = torch.ones([5, 9216, 64], dtype=torch.bfloat16, device=device)

y = foo(a, b)
print(y)
```
hlo:
```
HloModule SyncTensorsGraph.16, entry_computation_layout={()->(bf16[5,9216,64]{2,1,0})}

ENTRY SyncTensorsGraph.16 {
  constant.7 = bf16[] constant(1)
  reshape.8 = bf16[1,1,1]{2,1,0} reshape(constant.7)
  broadcast.9 = bf16[1,1,1]{2,1,0} broadcast(reshape.8), dimensions={0,1,2}
  reshape.10 = bf16[] reshape(broadcast.9)
  broadcast.11 = bf16[5,9216,64]{2,1,0} broadcast(reshape.10), dimensions={}
  convert.12 = f32[5,9216,64]{2,1,0} convert(broadcast.11)
  constant.1 = bf16[] constant(1)
  reshape.2 = bf16[1,1,1]{2,1,0} reshape(constant.1)
  broadcast.3 = bf16[1,1,1]{2,1,0} broadcast(reshape.2), dimensions={0,1,2}
  reshape.4 = bf16[] reshape(broadcast.3)
  broadcast.5 = bf16[5,9216,64]{2,1,0} broadcast(reshape.4), dimensions={}
  convert.6 = f32[5,9216,64]{2,1,0} convert(broadcast.5)
  multiply.13 = f32[5,9216,64]{2,1,0} multiply(convert.12, convert.6)
  convert.14 = bf16[5,9216,64]{2,1,0} convert(multiply.13)
  ROOT tuple.15 = (bf16[5,9216,64]{2,1,0}) tuple(convert.14)
} // SyncTensorsGraph.16
```

I was able to achieve bf16 multiplication by setting `export XLA_USE_BF16=1`, but I received the following warning
```
XLA_USE_BF16 will be deprecated after the 2.5 release, please convert your model to bf16 directly
```

I'm not sure how I can enable bf16 multiplication in HLO (High-Level Optimizer) in the correct way, without using the deprecated flag.

**Comment by avizon-aws** [comment]
I tried the same thing using autocast, and it seems to be working as you expect. Below is the code to replicate.
```
a = torch.ones([5, 9216, 64], dtype=torch.bfloat16, device=xm.xla_device())
b = torch.ones([5, 9216, 64], dtype=torch.bfloat16, device=xm.xla_device())

with torch.autocast(device_type='xla', dtype=torch.bfloat16):
    y=torch.mul(a, b)

xm.mark_step()
```

Below is the HLO:
```
HloModule SyncTensorsGraph.16, entry_computation_layout={()->(bf16[5,9216,64]{2,1,0}, bf16[5,9216,64]{2,1,0}, bf16[5,9216,64]{2,1,0})}

ENTRY %SyncTensorsGraph.16 () -> (bf16[5,9216,64], bf16[5,9216,64], bf16[5,9216,64]) {
  %constant.1 = bf16[] constant(1)
  %broadcast.5 = bf16[5,9216,64]{2,1,0} broadcast(bf16[] %constant.1), dimensions={}
  %constant.6 = bf16[] constant(1)
  %broadcast.10 = bf16[5,9216,64]{2,1,0} broadcast(bf16[] %constant.6), dimensions={}
  %constant.7 = bf16[] constant(1)
  %broadcast.12 = bf16[5,9216,64]{2,1,0} broadcast(bf16[] %constant.7), dimensions={}
  ROOT %tuple.15 = (bf16[5,9216,64]{2,1,0}, bf16[5,9216,64]{2,1,0}, bf16[5,9216,64]{2,1,0}) tuple(bf16[5,9216,64]{2,1,0} %broadcast.5, bf16[5,9216,64]{2,1,0} %broadcast.10, bf16[5,9216,64]{2,1,0} %broadcast.12), frontend_attributes={neff_output_names="output0,output1,output2"}
}

```

Can you try replicating and confirm.

**Comment by ysiraichi** [comment]
This is actually expected behavior. In fact, [PyTorch CUDA also does the same thing](https://github.com/pytorch/pytorch/blob/803017f3cb73bb115eda5ec0e0a19688ccafbf4e/aten/src/ATen/native/cuda/BinaryMulKernel.cu#L37-L42). 

In summary, PyTorch converts each operand (stored as `bf16`) to `at::OpMathType<bfloat16>` (i.e. `float`), runs the operation, and stores the result back as `bf16`. PyTorch/XLA just replicates the same behavior.

**Comment by avizon-aws** [comment]
To confirm` at::OpMathType<bfloat16>` this is still BF16 right? not float?

**Comment by avizon-aws** [comment]
I was not able to replicate the behavior which I posted previously. I still see an upcast of the inputs and the result is in FP32 which is then downcasted to BF16.

**Comment by ysiraichi** [comment]
Not really. [`at::OpMathType<bfloat16>::type == float`](https://github.com/pytorch/pytorch/blob/803017f3cb73bb115eda5ec0e0a19688ccafbf4e/aten/src/ATen/OpMathType.h#L23-L26). The same upcast-downcast also happens in PyTorch core.

**Comment by avizon-aws** [comment]
Thanks for confirming that its float and not BF16 thats used for upcast.

**Comment by apivovarov** [comment]
torch.mul is elemwise multiply.
Seems that nvidia GPUs before Hopper do not support BF16 elemwise multiply. XLA compiler upcasts ops such as Add, Multiply, Subtract to float32  in FloatNormalization pass (used in gpu_compiler.cc)
```
    case HloOpcode::kAdd:
    case HloOpcode::kMultiply:
    case HloOpcode::kSubtract: {
      if (LowPrecisionType() == BF16) {
               return cuda_compute_capability->IsAtLeastHopper();
      }
```
https://github.com/openxla/xla/blob/f24616501ce573c8320fe68e1604eb5f3c3cb473/xla/service/gpu/gpu_float_support.cc#L115

Maybe torch_xla or torch do the same operation legalization before XLA gpu_compiler.cc?


Ampere’s Tensor Cores support BF16 for matrix multiplication but not necessarily for element-wise ops.
Element-wise operations (e.g., multiply, add, subtract) are typically executed on CUDA cores, not Tensor Cores.
On Ampere, CUDA cores may not have optimized support for BF16 in general-purpose element-wise operations, which is why this restriction is enforced.
Conclusion:
On NVIDIA Ampere GPUs (like A100), XLA does NOT allow BF16 for element-wise multiply (HloOpcode::kMultiply).
On Hopper GPUs (like H100), XLA allows BF16 element-wise multiply.

**Comment by ysiraichi** [comment]
I'm not sure whether PyTorch does it due to GPU support, since the `OpMathType` for half is also float. PyTorch/XLA replicates this behavior by setting `use_opmathtype_for_compute()`:

https://github.com/pytorch/xla/blob/065cb5b1989dfe83af7d847c0c940181d37407d2/torch_xla/csrc/aten_xla_type.cpp#L2276-L2287

-----

> So is there any way to achieve pure BF16 multiplication without the flags?

One way to do that would be [to register a custom operation to PyTorch](https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html).

**Comment by miladm** [comment]
cc @tengyifei to assist

**Comment by tengyifei** [comment]
Based on @ysiraichi 's observations, this behavior sounds like working as intended?

**Comment by tengyifei** [comment]
Does anyone have the context as to:

- Why PyTorch chooses to upcast to f32
- Why we introduced the `XLA_USE_BF16` flag?

Without further context, it's hard to see what's the appropriate action here.

**Comment by ysiraichi** [comment]
> Why PyTorch chooses to upcast to f32

I was going to say that this is due to accuracy reasons, but I'm not so sure anymore. @amjames @ezyang Do you know why this is the case?

> Why we introduced the `XLA_USE_BF16` flag?

@miladm @zpcore @vanbasten23 @bhavya01 Do you know the reason?

**Comment by jeffhataws** [comment]
@JackCaoG could comment, but XLA_USE_BF16 and XLA_DOWNCAST_BF16 were originally introduced as a simple AMP, before AMP became available. Now that we have AMP, these flags are targeted for deprecation. We have deferred the full deletion for multiple releases to help with transition. Looks like this issue could be blocking the full deprecation?

**Comment by JackCaoG** [comment]
PyTorch/XLA has been pretty faithfully follow the pytorch behavior for the most part. However we do need to realize that pytorch is optimized for the GPU and TPU/other accelerator might have different capabilities. 

I don't think `XLA_USE_XXX` is a long term plan, it was an convenient hack back when mix precision/downcasting was difficult. If we think on TPU/XLA supported device it is better to do mul in bf16 we should change the default behavior(should be pretty easy).

**Comment by ysiraichi** [comment]
As @JackCaoG mentioned, it's an easy change. We could even modify the behavior based on the accelerator being used. It's just a matter of what makes sense for each accelerator. Let me know what you all think.

**Comment by avizon-aws** [comment]
For compatibility, we would want the support to be able to do multiplication in bf16 and fp32 based on the requirement, the flexibility to do both would be good. So if both inputs are in BF16, then its better to not promote types and perform the multiplication in BF16, else upcast to FP32 and proceed.

@JackCaoG could you share how to make these changes, also if I want to make these changes for a specific accelerator how can it be done?

**Comment by ysiraichi** [comment]
It's just a matter of calling `use_opmathtype_for_compute()`, there. You could condition it on the backend device, say call it only for XLA:CUDA.

**Comment by amjames** [comment]
In principal element-wise mul could be carried out in bf16, with matrix multiplication an upcast to f32 for the reduction has a justification based in maintaining accuracy. If this is a hardware limitation I am unaware of it and frankly would be a bit surprised as with the matrix multiply the multiplication is being done as bf16xbf16. 

The question here seems to be if pytorch/xla should addopt behaviors which make more sense for different accelerators. On that front I think the principal of least surprise would say that pytorch/xla should reproduce the behavior of pytorch core if the accelerator in question is supported by both, and otherwise follow the same guidelines unless there is a compelling justification to diverge behaviors.

**Comment by ysiraichi** [comment]
Issue opened inside PyTorch repo: https://github.com/pytorch/pytorch/issues/147134.

**Comment by miladm** [comment]
cc @lsy323 @ysiraichi to assist here.

**Comment by miladm** [comment]
@apivovarov @avizon-aws: as discussed offline, please share concrete specifics on the background of how this issue influences the XLA backend. Thank you.

**Comment by avizon-aws** [comment]
As suggested by @ysiraichi , a potential fix to ensure backwards compatibility is to comment this line.
https://github.com/pytorch/xla/blob/42edbe18bcd256869bc52e2562727c0f8516b2e6/torch_xla/csrc/aten_xla_type.cpp#L2284

It will ensure that the inputs are not upcasted to FP32. So, we can make this backend specific as well.

**Comment by ysiraichi** [comment]
I think it makes sense to make this behavior backend-specific. So, we would have:

- `XLA:CUDA`: upcast bf16 to fp32 in order to be consistent with PyTorch
- `XLA:TPU`: run the operation on the bf16 inputs

@miladm @avizon-aws @tengyifei @zpcore
Let me know what you all think.

**Comment by avizon-aws** [comment]
Yes, it sounds good to me, it will also make things more clear for the user, the upcasting behavior is not expected if inputs are in BF16 so the user doesnt know what is going on under the hood, unless he digs into the source code. I can make a PR for this to make it specific to backends e.g. TPU, Neuron. Will take a look into how to add this check.

**Comment by tengyifei** [comment]
Sgtm
